### PR TITLE
feat: add personal worth-it adjustment from listen history

### LINKED ISSUE #262: feat: Add personal worth-it adjustment from listen history
Part of #258

## Context

With per-episode topic tags (#259) and cross-episode topic rankings (#261) in place, the system has the data to answer the most valuable question for a busy professional: **"Is this episode worth MY time, given what I've already consumed?"**

Currently, the worth-it score is the same for every user. A 7.5-rated episode on "Leadership" is shown as 7.5 whether the user has listened to zero or five episodes on leadership. This misses the core value proposition: helping users be *selective* with their time.

This sub-issue adds a presentation-layer adjustment that uses the user's listen history and saved episodes to provide contextual "topic overlap" indicators. The stored score is never mutated — this is a per-user display overlay.

## Complexity Hint

**Subsystems touched:** Dashboard actions, recommendation queries, listen history queries, summary display components
**Independent stories:** 1
**Structural changes needed:** No — uses existing `listen_history`, `user_library`, and `episode_topics` tables; only adds query logic and UI indicators
**Scope note:** None detected

## Task

When displaying episode recommendations or scores, check the user's consumption history against the episode's topic tags and provide contextual indicators:

1. **Topic overlap detection:** Query user's listened/saved episodes → get their topic tags → compare with the current episode's topics
2. **Overlap indicator:** Show a contextual label when significant overlap exists, e.g.:
   - "You've heard 3 similar episodes" — high overlap, diminishing returns
   - "New topic for you" — no overlap, fresh content
   - "Top pick for [topic]" — best-ranked episode on a topic the user hasn't covered yet
3. **Recommendation ranking adjustment:** In the recommendations list, deprioritize episodes on topics the user has heavily consumed. Don't hide them — just sort them lower.

## Acceptance Criteria

- [ ] VERIFY: Given a user with listened episodes on topic "AI Ethics", when viewing an episode also tagged "AI Ethics", then a contextual overlap indicator is shown (e.g. "You've heard N similar episodes")
- [ ] VERIFY: Given a user with no listened episodes on topic "Leadership", when viewing a "Leadership" episode, then no overlap indicator is shown (or "New topic for you" is shown)
- [ ] VERIFY: Recommendations list deprioritizes episodes on topics the user has 3+ consumed episodes on
- [ ] VERIFY: Deprioritization is a sort adjustment, not a filter — overlapping episodes are still accessible
- [ ] VERIFY: Overlap detection considers both `listen_history` and `user_library` (saved episodes)
- [ ] VERIFY: Overlap calculation is efficient — batch query, not N+1
- [ ] VERIFY: Episodes with no topic tags (old episodes pre-#259) show no overlap indicator (graceful fallback)
- [ ] VERIFY: The stored `worth_it_score` is never modified — personal adjustment is display-only
- [ ] VERIFY: Unit tests cover: overlap detection logic, sort adjustment, edge cases
- [ ] VERIFY: `bun run lint`, `bun run test`, and `bun run build` pass

## Edge Cases

- User has no listen history or saved episodes — no indicators shown, recommendations use default sort
- Episode has no topic tags (pre-topic-extraction) — skip overlap check, no indicator
- User has consumed episodes on a topic but from different podcasts — still counts as overlap
- Episode belongs to multiple topics, user has overlap on only some — show overlap for the most relevant matching topic
- Very new user with 1-2 listened episodes — too little data for meaningful overlap, show indicators only when user has 3+ consumed episodes total

### Files to Modify

- `src/app/actions/dashboard.ts` — extend `getRecommendedEpisodes()` and `getRecentEpisodesFromSubscriptions()` to:
  1. Query user's consumed episode IDs from `listen_history` + `user_library`
  2. Get topic tags for consumed episodes from `episode_topics`
  3. Build a "user topic profile" (topic → count of consumed episodes)
  4. For each candidate episode, compute overlap score
  5. Add overlap metadata to returned DTO
  6. Adjust sort order: deprioritize episodes with high topic overlap
- `src/components/episodes/summary-display.tsx` — add overlap indicator below score display when overlap data is present
- `src/components/episodes/worth-it-badge.tsx` — compact overlap indicator for list views (e.g. small icon + tooltip)
- `src/components/podcasts/episode-card.tsx` — show overlap indicator if available in episode data

### Files to Create

- `src/lib/topic-overlap.ts` — pure function: given user topic profile + episode topics → overlap result `{ overlapCount: number, topOverlapTopic: string | null, isNewTopic: boolean }`

### Approach

1. Create `src/lib/topic-overlap.ts` with the overlap computation logic as a pure, testable function. Input: user topic profile `Map<string, number>`, episode topics `Array<{topic, relevance}>`. Output: overlap result.
2. In `dashboard.ts`, before returning recommendations:
   a. Batch-fetch the user's consumed episode IDs (listen_history + user_library)
   b. Batch-fetch topic tags for those episodes from `episode_topics`
   c. Build user topic profile: `Map<topic, consumedCount>`
   d. For each candidate episode, compute overlap using the pure function
   e. Attach overlap metadata to the returned DTO
   f. Re-sort: within same score tier, push high-overlap episodes lower
3. In `summary-display.tsx`, render overlap indicator when overlap data is present:
   - `overlapCount >= 3`: amber indicator "You've heard N similar episodes"
   - `overlapCount === 0`: green indicator "New topic for you" (only show if user has 5+ total consumed episodes — otherwise not meaningful)
   - Episode has `topicRank === 1` and `overlapCount === 0`: highlight as "Top pick for [topic]"
4. In `worth-it-badge.tsx` and `episode-card.tsx`, add compact overlap indicator (icon or small label).
5. Write unit tests for overlap computation and sort adjustment.
6. Test end-to-end with a user who has listen history.

### Patterns to Follow

- Follow `src/app/actions/dashboard.ts` `getRecommendedEpisodes()` query pattern for batch fetching.
- Follow `src/lib/score-utils.ts` for utility function style (pure functions, no side effects).
- Follow existing component patterns in `summary-display.tsx` for conditional rendering of indicators.

### Constraints

- **Performance:** Overlap computation must not add noticeable latency to dashboard load. Batch all queries, avoid N+1. Target: <50ms additional query time.
- **No score mutation:** The stored `worth_it_score` column must never be modified by this feature. Overlap is a display-layer concept.
- **Progressive enhancement:** If topic tags don't exist for an episode (old data), the feature gracefully degrades — no indicators, no sort adjustment for that episode.

### In Scope
- Topic overlap detection from listen history + saved episodes
- Contextual indicators in episode display (summary, badge, card)
- Recommendation sort adjustment based on topic overlap
- Pure overlap computation function with unit tests

### Out of Scope
- Personalized scoring model (ML-based, collaborative filtering) — future iteration
- User preference for topic interests (explicit topic follows/mutes)
- Notifications about new episodes on "new" topics for the user
- Email/push digest with personalized picks

## Dependencies

- Blocked by: #259 (topic tags), #261 (topic ranking — for "Top pick" badges)
- Blocks: None

## Definition of Done

- [ ] Implementation complete
- [ ] Tests passing (unit + integration)
- [ ] `bun run lint`, `bun run test`, `bun run build` pass
- [ ] PR approved and merged
- [ ] Epic checklist updated

### PR DESCRIPTION
**Pull request by funny-otter**
## Summary

- Adds display-only topic overlap indicators that personalize episode recommendations using the user's consumption history (listen history + saved episodes)
- Shows contextual labels: "You've heard N similar episodes" (amber), "New topic for you" (green), "Top pick for [topic]" (green) — gated by a 3-episode minimum
- Deprioritizes heavily-consumed topics in recommendations via stable partition sort (sort adjustment, not filter)
- Pure function pattern (`src/lib/topic-overlap.ts`) with batch queries (no N+1), stored `worth_it_score` is never modified

### Architecture
- **New**: `src/lib/topic-overlap.ts` — pure overlap computation (24 unit tests)
- **New**: `docs/adr/034-personal-topic-overlap-indicators.md`
- **Modified**: `dashboard.ts` — user topic profile construction, overlap enrichment, sort adjustment, `getEpisodeTopicOverlap()` action
- **Modified**: `episode-recommendations.tsx`, `summary-display.tsx`, episode page — overlap label rendering

Closes #262

## Test plan

- [x] `bun run lint` — no warnings or errors
- [x] `bun run test` — 1620/1620 tests pass (132 files), including 24 new overlap unit tests and 16 new dashboard integration tests
- [x] `bun run build` — compiles successfully
- [ ] Manual: verify overlap indicators appear for users with listen history on recommendations
- [ ] Manual: verify episode detail page shows overlap label via non-blocking useEffect
- [ ] Manual: verify users with <3 consumed episodes see no indicators
- [ ] Manual: verify episodes without topic tags show no indicator

## Summary by CodeRabbit

* **New Features**
  * Personalized topic-overlap indicators added to recommendation cards and episode summary; labels highlight high-overlap, top-pick, or new-topic status.
  * Recommendation lists are stably reordered to deprioritize heavily-consumed topics.

* **Documentation**
  * Added design/ADR and changelog entries describing the display-only indicator behavior and gating rules.

* **Tests**
  * New unit and integration tests covering overlap computation, labeling rules, ordering, and graceful fallback.

### CODE REVIEW

File: src/app/actions/dashboard.ts [hunk 1]
```diff
@@ -267,13 +272,169 @@ export async function getRecommendedEpisodes(
...
+        candidateTopicRows = await db
+          .select({
+            episodeId: episodeTopics.episodeId,
+            topic: episodeTopics.topic,
+            relevance: episodeTopics.relevance,
+          })
+          .from(episodeTopics)
+          .where(inArray(episodeTopics.episodeId, candidateIds));
+      }
+
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/app/actions/dashboard.ts:325]: Fixed: added `.orderBy(desc(episodeTopics.relevance))` to candidate topics query for deterministic topic ordering.

File: src/app/actions/dashboard.ts [hunk 2]
```diff
@@ -267,13 +272,169 @@ export async function getRecommendedEpisodes(
...
+    // Fetch topics and best rank for this episode
+    const epTopicRows = await db
+      .select({
+        topic: episodeTopics.topic,
+        relevance: episodeTopics.relevance,
+        topicRank: episodeTopics.topicRank,
+      })
+      .from(episodeTopics)
+      .where(eq(episodeTopics.episodeId, episodeDbId));
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/app/actions/dashboard.ts]: Fixed: added `.orderBy(desc(episodeTopics.relevance))` to episode topics query.

File: src/app/(app)/episode/[id]/page.tsx [hunk 3]
```diff
@@ -321,6 +323,15 @@ export default function EpisodePage({ params }: EpisodePageProps) {
...
   }, [isOnline, fetchEpisodeData, loadFromCache]);
 
+  // Non-blocking: fetch topic overlap indicator after episode data loads.
+  // Fires only when online and episode data is present.
+  useEffect(() => {
+    if (!isOnline || !episode) return;
+    void getEpisodeTopicOverlap(episodeId).then((result) => {
+      setOverlapLabel(result.label);
+    });
+  }, [isOnline, episode, episodeId]);
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/app/(app)/episode/[id]/page.tsx]: Fixed: already addressed in prior commit — useEffect now uses `ignore` flag pattern with merged state and boolean dependency.

File: docs/adr/034-personal-topic-overlap-indicators.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+   * e.g., if user consumed 5 "AI" episodes and 2 "Leadership" episodes,
+   * and the current episode is tagged ["AI", "Leadership"],
+   * overlapCount = 5 (the max), topOverlapTopic = "AI".
+   */
+  overlapCount: number;
+  topOverlapTopic: string | null;  // the topic with highest overlap
+  isNewTopic: boolean;         // true if ALL episode topics are new to the user
+  label: string | null;        // pre-computed display label, null if no indicator
+}
+```
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply docs/adr/034-personal-topic-overlap-indicators.md:50]: Fixed: added `labelKind` discriminator field to ADR interface definition.

File: docs/adr/034-personal-topic-overlap-indicators.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+### DTO extension
+
+`RecommendedEpisodeDTO` gains optional overlap fields:
+```typescript
+overlapCount?: number;
+overlapTopic?: string | null;
+overlapLabel?: string | null;
+```
+
+These are optional so existing consumers (tests, other call sites) don't break. The `RecentEpisode` type from the PodcastIndex feed does not get overlap — those episodes come from an external API and may not have topic tags in our DB. The issue focuses on recommendations and the episode detail page.
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply docs/adr/034-personal-topic-overlap-indicators.md:78]: Fixed: added `overlapLabelKind` field to ADR DTO extension section.

**Comment by funny-otter** [comment]
> 🧹 Nitpick comments (2) — ADR wording and duplicated label-color logic...

Fixed: Both nitpicks already addressed in prior commits. ADR line 84 correctly states "inside the Worth-It Score card." Label color logic uses `overlapLabelKind` discriminator instead of `startsWith()` string matching.
